### PR TITLE
ENH: Move default [con/de]strutor to .h (part 2)

### PR DESCRIPTION
**Pull request by hjmjohnson**
Use C++11 = default in .h file rather than
definition in .hxx.

This work is a continuation of work started in Core/Common
subdirectory, but now is applied to all the non-third party
directories

Skipping a few items that need further evaluation:
  Modules/IO/TransformBase*
  Modules/Numerics/Optimizersv4/include/itkObjectToObjectOptimizerBase.h

Co-authored-by: Mathew J Seng <mathew-seng@uiowa.edu>

### CODE REVIEW

**Comment by blowekamp** [comment]
I am not in favor of removing detractors form the compiled libraries because of increased object file size, and potential dynamic_cast/duplicate symbol issues.

**Comment by hjmjohnson** [comment]
@blowekamp I don't this that we are changing any of the behaviors with regards explicit instantiations.

It is my understanding that having 

`~myclass() = default;` in the .h file

is equivalent to having

`my class::~myclass()` = default; in the .hxx (which is included at the end of header file) file

The explicit instantiations cases are currently excluded form this patch set, and that *may* be a case where there is a difference.

Can you elaborate on the scenario where increased object size would occur?

File: Modules/Core/Common/src/itkLightProcessObject.cxx [hunk 1]
```diff
@@ -29,13 +29,6 @@ ::LightProcessObject()
   m_Progress = 0.0f;
 }
 
-/**
- * Destructor for the LightProcessObject class. We've got to
- * UnRegister() the use of any input classes.
- */
-LightProcessObject
-::~LightProcessObject() = default;
-
```
Comment (blowekamp) [inline Modules/Core/Common/src/itkLightProcessObject.cxx]: There are several *.cxx files which have the default destructor removed from them. It is my understanding that moving this method to be implicitly inline in the header, will create an instance of the method in every compilation unit. Where as with the prior "opaque" declaration in the header the library instance would have to be used.

**Comment by hjmjohnson** [comment]
Ahh!  That I can see *might* make a difference, and to be safe I'll remove those for now as this was intended to be only stye changes.

But there are other considerations as well:

https://stackoverflow.com/questions/38242200/where-should-a-default-destructor-c11-style-go-header-or-cpp

Use non-user defined destructors can allow for additonal compile time optimizations.

**Comment by blowekamp** [comment]
> Use non-user defined destructors can allow for additonal compile time optimizations.

I believe for these cases the the automatic `=default` or non-user defined implementation is still used. It is just compiled in the library, as opposed to being inlined in the header. 

But inlined implementations can provide opportunities for additional optimizations at the potential trade-offs increased code size and compilation times. As one trying to insatiate much of ITK for binary libraries the later is a significant concern to me. However, there is also the [IPO](https://en.wikipedia.org/wiki/Interprocedural_optimization) compile options which may return some inline benefits when not explicitly coded that way.

**Pull request merged**